### ISSUE TITLE
ec2_asg rolling replace creates extra, undesired instance

### ISSUE THREAD

**Issue by ansibot** [ansible/ansible#30075]
##### ISSUE TYPE

Bug Report

##### COMPONENT NAME

ec2_asg module

##### ANSIBLE VERSION

1.9

##### SUMMARY

From @ferrants on 2015-07-30T19:08:37Z
Experienced with `stable-1.9`.

Desired size is incremented by the batch size. The old instances are marked as unhealthy. As the old instance is terminated, a new instance is created to match the desired size. Once the old instance is terminated and all new instances are healthy, the size is decreased, terminating undesired instance.

With a batch size of 1 and desired size of one, the flow is as follows:
- starts with one old instance, desired is 1
- gets one new instance, desired is 2
- old instance starts being torn down, gets another new instance, desired is 2
- one of the new instances is torn down when desired capacity is reduced to 1

Activity history is attached:
<img width="1207" alt="screen shot 2015-07-30 at 3 08 09 pm" src="https://cloud.githubusercontent.com/assets/882183/8992509/d50739aa-36cc-11e5-9681-eb4c47615636.png">

Copied from original issue: ansible/ansible-modules-core#1845

**Comment by ansibot** [comment]
From @dainiax on 2015-07-30T19:08:37Z

Got the same bug (ansible 1.9.2)
from task:
...
- name: Create/edit auto-scaling group ASG<br />
  ec2_asg:<br />
  aws_access_key: "{{ AWS_ACCESS_KEY }}"<br />
  aws_secret_key: "{{ AWS_SECRET_KEY }}"<br />
  name: ASG_NAME<br />
  launch_config_name: "{{ NEW_LC.name }}"<br />
  health_check_period: 120<br />
  health_check_type: EC2<br />
  replace_all_instances: true<br />
  replace_batch_size: 2<br />
  min_size: 2<br />
  max_size: 2<br />
  desired_capacity: 2<br />
  wait_for_instances: true<br />
  region: "{{ AWS_REGION }}"<br />
  register: EC2_ASG_INFO<br />
  ...

And in CloudTrail on the same time for this task I see events
...
"minSize": 2,
"desiredCapacity": 2,
"autoScalingGroupName": "ASG_NAME",
"defaultCooldown": 300,
"healthCheckType": "EC2",
"maxSize": 2,
...
and right after, another event on the same time
...
"minSize": 4,
"desiredCapacity": 4,
"autoScalingGroupName": "ASG_NAME",
"defaultCooldown": 300,
"healthCheckType": "EC2",
"maxSize": 4
...

**Comment by ansibot** [comment]
From @ferrants on 2015-07-30T19:08:37Z

@bcoca - do you know if anyone is looking into this? I may be able to spend time digging in and producing a fix if not. I don't want to duplicate efforts.

**Comment by ansibot** [comment]
From @ferrants on 2015-07-30T19:08:37Z

@dainiax - you have not described the same issue, with a replace size of 2, it should be at a capacity of 4 for a while, but 2 should be of the old launch config and 2 of the new launch config. The issue is that there are 4 instances of the new launch config.

**Comment by ansibot** [comment]
From @dainiax on 2015-07-30T19:08:37Z

@ferrants - the same problem, maybe I wasn't sure whats really going on. So run few more tests.
1. Had LC_a (Launch Config), ASG_a (Auto-scaling group) with params min/max/desired size of 1 and one instance running, lets name inst_a.
2. Launched Ansible script to create new LC_b and update ASG_a (same sizes min/max/desired 1, replace size was not specified).
3. New instance, lets say inst_b was launched.
4. Inst_a was torn down.
5. ! Another new instance was launched inst_c.
6. ! Inst_b was torn down.
7. Left running single inst_c
Result was what I wanted, but in process redundant new instance was started and later torn down.
And I think it depends on parameter "replace_all_instances" of how mane redundant instances will be launched and shut down.

I don't think anyone is looking into this, as I googled a lot about this issue. And its also not my expertise to fix this. So if you could try to fix it, would be really nice.

**Comment by ansibot** [comment]
From @robbwagoner on 2015-07-30T19:08:37Z

I see this happening with a `batch_size` of 2, as well. It doesn't always happen, though. I.e. I can reproduce consistently.
 However, it does manifest itself with ASG's `min_size`, `max_size`, and `desired` eventually increasing by `batch_size`. Normally our ASG is defined with `min_size` of 2 and `max_size` of 8. Now they are 4 and 10, respectively. We relaunch/refresh the ASG 3 times a day and it seems to happen after a few days. 

I've looked at the module code. The workflow and logic is, admittedly, complicated. If I had the time available, I'd work on a fix. But, I cannot commit to doing so at this time. I'd be happy to test out a development branch, though.

**Comment by ansibot** [comment]
From @ferrants on 2015-07-30T19:08:37Z

@robbwagoner, that is expected with the rolling replace. It increases the capacity by the `batch_size` so it scales up and ensures everything is healthy before marking old instances as unhealthy.

The issue here is that a new, healthy instance will be terminated during the rolling deploy. So, it rolls too many times because the ASG size is too high for just a bit too long. It needs to specify the correct `desired` at the same time (or maybe before) the last instance is marked as unhealthy so when it scales down, it is at the right size and no additional scaling occurs.

**Comment by ansibot** [comment]
From @JadyAllure on 2015-07-30T19:08:37Z

@ferrants we met exactly the same issue when use ec2_asg module to implement rolling update. Ansible is trying to terminate instances with old launch configuration before decrease the size of asg. 
AWS responses faster to launch new extra instances to meet the asg size.

**Comment by ansibot** [comment]
From @ferrants on 2015-07-30T19:08:37Z

I've been meaning to implement and submit a fix but haven't found the time. I'll probably upgrade to ansible 2.0 and re-assess before attempting a fix. I'm currently using our internal fork https://github.com/dataxu/ansible-modules-core/tree/dx-stable-1.9

**Comment by ansibot** [comment]
From @btd on 2015-07-30T19:08:37Z

Got the same bug. Would like to know if anybody working on fix.

**Comment by ansibot** [comment]
From @JadyAllure on 2015-07-30T19:08:37Z

Personally, I have two recommandations.
1. trying with Ansible2 - I was on Ansible1.9x when having this problem
2. we decided to use ansible+cloudformation rolling update instead of purely Ansible deployment. This works a treat.

Please read this very useful article from Ansible - http://www.ansible.com/blog/immutable-systems

**Comment by ansibot** [comment]
From @wholroyd on 2015-07-30T19:08:37Z

I'm using Ansible 2.0 and having this issue with max:1/min:0/desired:1 on ASG when changing the LC.

I'll temporarily have two new instances, but then only one ends up surviving after a minute or two.

**Comment by ansibot** [comment]
From @ferrants on 2015-07-30T19:08:37Z

I'm still experiencing this issue as well. I plan on using ansible+cloudformation in the near-future.

**Comment by ansibot** [comment]
From @wholroyd on 2015-07-30T19:08:37Z

Looking at the module code, it seems to update the ASG size requests (current sizes + batch size) so that it can spin up new instances in addition to the old instances, then kill off each old instance (per batch size) at a time, until the new instance size is reached, then exits and drops the ASG sizes back down.

It seems like it's not able to do this in time before the ASG realizes desired count isn't met and ends up starting up an instance automatically at the same time the module requests a new one itself. It also appears to tear down active instances before any new ones come online first, so you could end up with an empty ASG before a new one is capable of taking traffic.

I think removing the adjustment to desired and min count from the upsizing of the ASG should help prevent this. I'm already in the module looking at #3160, so I'll poke at it.

**Comment by ansibot** [comment]
From @wholroyd on 2015-07-30T19:08:37Z

Looks like one other person verified my fix in #3162, just waiting on the core team to review it.

**Comment by ansibot** [comment]
From @ferrants on 2015-07-30T19:08:37Z

What does @ansibot do if it gets ignored forever?

**Comment by ansibot** [comment]
From @jctanner on 2015-07-30T19:08:37Z

@ferrants nothing at the moment.

**Comment by ansibot** [comment]
From @morganchristiansson on 2015-07-30T19:08:37Z

I'm having this issue. Is this issue being tracked anywhere else?

**Comment by ansibot** [comment]
@ansibot Greetings! Thanks for taking the time to open this issue. In order for the community to handle your issue effectively, we need a bit more information. 

Here are the items we could not find in your description:
- issue type
- ansible version
- component name

Please set the description of this issue with this template:
https://raw.githubusercontent.com/ansible/ansible/devel/.github/ISSUE_TEMPLATE.md

[click here for bot help](https://github.com/ansible/ansibullbot/blob/master/ISSUE_HELP.md)

**Comment by biomassives** [comment]
This message was created automatically by mail delivery software.

A message that you sent could not be delivered to one or more of its
recipients. This is a temporary error. The following address(es) deferred:

  acmeideal@gmail.com
    Domain biomassiv.es has exceeded the max emails per hour (123/100 (123%)) allowed.  Message will be reattempted later

------- This is a copy of the message, including all the headers. ------
Received: from github-smtp2-ext1.iad.github.net ([192.30.252.192]:46246 helo=github-smtp2a-ext-cp1-prd.iad.github.net)
	by chi-server32.websitehostserver.net with esmtps (TLSv1.2:ECDHE-RSA-AES256-GCM-SHA384:256)
	(Exim 4.89)
	(envelope-from <noreply@github.com>)
	id 1drnHz-000kQj-0R
	for greg@biomassiv.es; Tue, 12 Sep 2017 10:39:51 -0500
Date: Tue, 12 Sep 2017 08:33:51 -0700
DKIM-Signature: v=1; a=rsa-sha256; c=relaxed/relaxed; d=github.com;
	s=pf2014; t=1505230750;
	bh=X/9BwMeDJxfwbvDvKwNdQ+27iRUagajCgSq2HNUmIQ8=;
	h=From:Reply-To:To:Cc:In-Reply-To:References:Subject:List-ID:
	 List-Archive:List-Post:List-Unsubscribe:From;
	b=zgoufLoBx+18Usi4W0S5ulK3GoaXaxrzPpbld0hxO+Xs+hujrIMAU0STW9B3saX/y
	 KA1En7ZF2xzn8i5L1SRl6C/XoVed+i+NxmrO03bnA7rEqoUY+W97FmOoD5ObqJD/B/
	 jpqT9AnoAtb6nL1dU5RckUXOJFDeSDVmFPxoJUF0=

**Comment by biomassives** [comment]
This message was created automatically by mail delivery software.

A message that you sent could not be delivered to one or more of its
recipients. This is a temporary error. The following address(es) deferred:

  acmeideal@gmail.com
    Domain biomassiv.es has exceeded the max emails per hour (124/100 (124%)) allowed.  Message will be reattempted later

------- This is a copy of the message, including all the headers. ------
Received: from github-smtp2-ext5.iad.github.net ([192.30.252.196]:53038 helo=github-smtp2a-ext-cp1-prd.iad.github.net)
	by chi-server32.websitehostserver.net with esmtps (TLSv1.2:ECDHE-RSA-AES256-GCM-SHA384:256)
	(Exim 4.89)
	(envelope-from <noreply@github.com>)
	id 1drnIE-000kXu-4A
	for greg@biomassiv.es; Tue, 12 Sep 2017 10:40:06 -0500
Date: Tue, 12 Sep 2017 08:34:14 -0700
DKIM-Signature: v=1; a=rsa-sha256; c=relaxed/relaxed; d=github.com;
	s=pf2014; t=1505230765;
	bh=/CMq7r38KXkLyPscQIRm39wSb/cOjwv0JCJXuxjd+UA=;
	h=From:Reply-To:To:Cc:In-Reply-To:References:Subject:List-ID:
	 List-Archive:List-Post:List-Unsubscribe:From;
	b=js1HfsgMik85Em8zspLzn8z/ia3MGPq7AwdOdydBS9u3HO+88+6pEBYC8ZUQvlarv
	 v/1V253LPkZe2KKkjMu0Oz2r/pP9/QMOjDXq47gflv2IK78gcwu5FoZlO1++BVU2xV
	 Lgsg5ppfX+L6liLK+F59F4SPF3KJfFcyyFsMx5kQ=

**Comment by ansibot** [comment]
cc @garethr
[click here for bot help](https://github.com/ansible/ansibullbot/blob/master/ISSUE_HELP.md)

**Comment by sylvinus** [comment]
@jctanner I'm having this issue with 2.3 as well so I don't think it's specific to 1.9

**Comment by ansibot** [comment]
cc @ryansb @s-hertel
[click here for bot help](https://github.com/ansible/ansibullbot/blob/master/ISSUE_HELP.md)

**Comment by ansibot** [comment]
cc @willthames
[click here for bot help](https://github.com/ansible/ansibullbot/blob/master/ISSUE_HELP.md)

**Comment by IronCore864** [comment]
Having this issue with 2.4, when min/max/desired are both 1.

**Comment by ansibot** [comment]
cc @jillr @tremble
[click here for bot help](https://github.com/ansible/ansibullbot/blob/master/ISSUE_HELP.md)

**Comment by anonfriese** [comment]
fyi still having this issue with 2.9.5....

**Comment by ansibot** [comment]
Thank you very much for your interest in Ansible. Ansible has migrated much of the content into separate repositories to allow for more rapid, independent development. We are closing this issue/PR because this content has been moved to one or more collection repositories.

* lib/ansible/modules/cloud/amazon/ec2_asg.py -> https://galaxy.ansible.com/community/aws

For further information, please see:
https://github.com/ansible/ansibullbot/blob/master/docs/collection_migration.md
